Added example of using libsigc++ with Qt

rm5248 opened the pull request.

murraycu commented:
Thanks, but what is this example showing us?

Is it showing us how to call a libsigc++ slot from a Qt signal handler (slot)? If so, is that particularly difficult to figure out, or difficult to do when also using Qt? Isn't it just like calling a function?

Or is it just an example of how you need to disable the Qt "keywords", instead using the Q_* macros?

rm5248 commented:
> Is it showing us how to call a libsigc++ slot from a Qt signal handler (slot)? If so, is that particularly difficult to figure out, or difficult to do when also using Qt? Isn't it just like calling a function?

There's not a big difference between the Qt signals/slots and libsigc++, but they can be used in a similar manner.

> Or is it just an example of how you need to disable the Qt "keywords", instead using the Q_* macros?

This is the needed part.  I also wanted to show that you can use both the Qt signals/slots at the same time as libsigc++.

Because of that, it is a relatively trivial example, as otherwise it could get very complicated.

murraycu commented:
OK. Thanks. Then I'd like it to be clearer about that, please. I think it should be as minimal as possible - it should have Qt signals and slots and libsigc++ signals and slots, just to show how they can co-exist within the same class (because of the use of the Q_* keywords) but they should not interact.

And I think there should be a README.md saying what it is showing.

rm5248 commented:
OK, I've added a README.md and simplified the example slightly.  The signals don't interact with one another much; the slot that is used for sigc++ is also a Qt slot, but I feel that it is important to mention.  The one thing that the example does that is not a good practice is emitting signals inside the constructor; would it make sense to have that in a separate method for clarity?

murraycu commented on examples/qt/README.md:
```diff
@@ -0,0 +1,45 @@
+happen correctly.
+
+Steps to use libsigc++ with Qt:
+1. In your .pro file, add `CONFIG += no_keywords`.  This configures Qt to not
```
This assumes that the person is using qmake, I think. But Qt doesn't require the use of qmake.

murraycu replied:
Also, I think it's worth saying that this (using the Q_* macros) is generally good advice when writing Qt code.

rm5248 replied:
I'm generally in agreement with that, I'll go update this in a bit.  While I don't disagree that using the Q_* macros is a good idea, that's generally not what happens in most Qt code that I have seen(even Qt's own examples use `emit`/`signals`/`slots` and not the Q_* macros).

murraycu commented on examples/qt/README.md:
```diff
@@ -0,0 +1,45 @@
+
+Steps to use libsigc++ with Qt:
+1. In your .pro file, add `CONFIG += no_keywords`.  This configures Qt to not
+define the keywords `emit`, `signals`, and `slot`.
```
These are macros, not keywords.

rm5248 replied:
They're macros in the C++ land, but that is paraphrasing what the Qt documentation says in the link you put above: 

> It tells Qt not to define the moc keywords `signals`, `slots`, and `emit` [...]

rm5248 commented:
Updated based off of feedback.

murraycu commented:
Many thanks, and sorry for being so slow to respond.

I have squashed these into one commit and pushed them to master.

This directory is not DISTed into the tarball (by make distcheck), and we should fix that, but it's not fatal.

kjellahl commented:
>This directory is not DISTed into the tarball (by make distcheck)

It's not distributed by `make distcheck`. It *is* distributed by `ninja dist`.
Meson/ninja distributes all files from the git repo, unless instructions in a
meson.build file removes some files from the meson-dist directory.

The example with Qt is not built by a meson build or an autotools build. If it's
ever added to the build systems, building it should be controlled by a configuration
option, and disabled by default, like the benchmark with Boost.
